File: store/jsp/src/store/ProductDetails.java
package store;

import store.cart_item;
import store.Product;
import store.Mysql;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletResponse;
import javax.servlet.http.HttpSession;
import javax.servlet.ServletException;
import java.util.ArrayList;
import java.util.TreeMap;
import java.io.IOException;
import java.sql.ResultSet;
import java.sql.Connection;
import java.sql.SQLException;

public class ProductDetails extends HttpServlet {
    private Connection connection;

    void add_to_cart(HttpServletRequest request, HttpServletResponse response) throws IOException, SQLException {
        HttpSession session = request.getSession(false);
        if(session==null) throw new RuntimeException("no session");
        ArrayList cart = (ArrayList) session.getAttribute("cart");
        if(cart == null)
            cart = new ArrayList();
        cart_item item = new cart_item();
        item.quantity = Integer.parseInt(request.getParameter("quantity"));
        item.product = fetch(request.getParameter("product_id"));
        if(item.quantity <= 0)
            throw new RuntimeException("Quantity is invalid. It must be a positive number");
        cart.add(item);

        session.setAttribute("cart",cart);
        response.sendRedirect("ShoppingCart.sxx");
    }
    Product fetch(final String id) throws SQLException {
        final String sql = "SELECT * FROM store_products WHERE product_id=$id";
        TreeMap params = new TreeMap();
        params.put("id",id);

        ResultSet result = Mysql.query(connection, Mysql.prepare(sql, params));
        if( result.next() ) {
            Product product = new Product();
            product.product_id = Long.parseLong(result.getString(1));
            product.category_id = Long.parseLong(result.getString(2));
            product.name = result.getString(3);
            product.price = Float.parseFloat(result.getString(4));
            product.img = result.getString(5);
            product.desc = result.getString(6);
            product.recomended = Boolean.getBoolean(result.getString(7));
            return product;
        }else
            throw new RuntimeException("product_id "+id+" not found");
    }
    void show_product(HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException, SQLException {
        String id = request.getParameter("product_id");
        if(id==null)
            throw new RuntimeException("ProductDetails: product_id parameter not supplyed");
        Product product = fetch(id);
        request.setAttribute("product", product);
        request.getRequestDispatcher("ProductDetailsView.jsp").forward(request,response);
    }
    public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException {
        try{
            try{
                add_to_cart(request, response);
            }catch(Throwable ex) {
                ex.printStackTrace();
                request.setAttribute("error",ex.getMessage());
                show_product(request, response);
            }
        }catch(Throwable ex) {
            ex.printStackTrace();
            request.setAttribute("error",ex.getMessage());
            request.setAttribute("product",new Product());
            request.getRequestDispatcher("ProductDetailsView.jsp").forward(request,response);
        }
    }
/*

    public void init() throws ServletException {
        connection = Mysql.create();
    }

    public void destroy() {
        Mysql.destroy(connection);
    }
*/

    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException {
        connection = Mysql.get();
        try{
            request.getSession(true); //start session
            show_product(request, response);
        }catch(Throwable ex) {
            ex.printStackTrace();
            request.setAttribute("error",ex.getMessage());
            request.setAttribute("product",new Product());
            request.getRequestDispatcher("ProductDetailsView.jsp").forward(request,response);
        }
    }
};

File: store/cxxsp/ProductDetails.cpp
#include <servlet/HttpServlet.h>
#include <servlet/RequestDispatcher.h>
#include <boost/serialization/vector.hpp>
#include <stdexcept>
#include <unistd.h>
#include <stdlib.h>
#include <iostream>
#include <sstream>
#include <mysql/mysql.h>

#include <algorithm>
#include <vector>
#include <boost/lexical_cast.hpp>
#include <boost/shared_ptr.hpp>
#include "cart_item.h"
#include "Product.h"
#include "mysql.h"

using namespace std;
using namespace servlet;

using boost::lexical_cast;
using boost::shared_ptr;

class ProductDetails : public HttpServlet {
    void add_to_cart(HttpServletRequest& request, HttpServletResponse &response);
    Product fetch(const std::string& id);
    void show_product(HttpServletRequest& request, HttpServletResponse &response);
public:
    void doPost(HttpServletRequest& request, HttpServletResponse &response);
    void doGet(HttpServletRequest& request, HttpServletResponse &response);
};

//-------------------------------------------------------------------------------

void ProductDetails::add_to_cart(HttpServletRequest& request, HttpServletResponse &response) {
    HttpSession* session = request.getSession(false);
    if(!session) throw runtime_error("no session");
    typedef vector<cart_item> shopping_cart;
    shopping_cart cart = session->getAttribute<shopping_cart>("cart");
    cart_item item;
    item.quantity = boost::lexical_cast<int>(request.getParameter("quantity"));
    item.product = fetch(request.getParameter("product_id"));
    if(item.quantity <= 0)
        throw runtime_error("Quantity is invalid. It must be a positive number");
    cart.push_back(item);

    session->setAttribute("cart",cart);
    response.sendRedirect("ShoppingCart.sxx");
}

Product ProductDetails::fetch(const std::string& id) {
    const string sql = "SELECT * FROM store_products WHERE product_id=$id";
    map<string,string> params;
    params["id"] = id;

    shared_ptr<MYSQL_RES> result = mysql.query(mysql.prepare(sql, params));
    MYSQL_ROW row;
    if( (row = mysql_fetch_row(result.get())) ) {
        Product product;
        product.product_id = lexical_cast<long>(mysql.get(row,0));
        product.category_id = lexical_cast<long>(mysql.get(row,1));
        product.name = mysql.get(row,2);
        product.price = lexical_cast<float>(mysql.get(row,3));
        product.img = mysql.get(row,4);
        product.desc = mysql.get(row,5);
        product.recomended = lexical_cast<bool>(mysql.get(row,6));
        return product;
    }else
        throw runtime_error("product_id "+id+" not found");
}

void ProductDetails::show_product(HttpServletRequest& request, HttpServletResponse &response) {
    string id = request.getParameter("product_id");
    if(id.empty())
        throw runtime_error("ProductDetails: product_id parameter not supplyed");
    Product product = fetch(id);
    request.setAttribute("product", product);
    request.getRequestDispatcher("ProductDetailsView.csp")->forward(request,response);
}

void ProductDetails::doPost(HttpServletRequest& request, HttpServletResponse &response) {
    try{
        try{
            add_to_cart(request, response);
        }catch(const exception& ex) {
            request.setAttribute<string>("error",ex.what());
            show_product(request, response);
        }
    }catch(const exception& ex) {
        request.setAttribute<string>("error",ex.what());
        request.setAttribute("product",Product());
        request.getRequestDispatcher("ProductDetailsView.csp")->forward(request,response);
    }
}

void ProductDetails::doGet(HttpServletRequest& request, HttpServletResponse &response) {
    try{
        request.getSession(true); //start session
        show_product(request, response);
    }catch(const exception& ex) {
        request.setAttribute<string>("error",ex.what());
        request.setAttribute("product",Product());
        request.getRequestDispatcher("ProductDetailsView.csp")->forward(request,response);
    }
}


EXPORT_SERVLET(ProductDetails);
